client/securedrop_client/gui/base/progress.py

```diff
@@ -0,0 +1,102 @@
+import math
+import time
+
+from PyQt5.QtCore import QTimer, pyqtSignal
+from PyQt5.QtWidgets import QProgressBar
+
+from securedrop_client.sdk import ProgressProxy
+from securedrop_client.utils import humanize_speed
+
+SMOOTHING_FACTOR = 0.3
+
+
+class FileDownloadProgressBar(QProgressBar):
+    """
+    A progress bar for file downloads.
+
+    It receives progress updates from the SDK and updates the total % downloaded,
+    as well as calculating the current speed.
+
+    We use an exponential moving average to smooth out the speed as suggested by
+    https://stackoverflow.com/a/3841706; the reported speed is 30% of the current
+    speed and 70% of the previous speed. It is updated every 100ms.
+    """
+
+    # One of:
+    # {"size": int}
+    # {"decrypting": True}
+    signal = pyqtSignal(dict)
```

Comment:
Yeah, same. That's what I originally had but switched it to a dict to accommodate the decrypting signal.

> I wonder if it may be simpler to have "decrypting" ui behaviour fully separate from the download progress bar (eg: download progress is finished -> hide the progress bar and show a widget relevant to decryption).

Do you think that other widget would be something other than a progress bar? Once we switch to Sequoia I expect we'll be able to calculate decryption progress just like we can with downloads now. At that point I imagine we could do something like the first 90% of the progress bar is download and the last 10% is decrypting/unzippping. Or just two progress bars, but either way I'd imagine it's the same underlying QProgressBar widget.

Certainly we can split it into two signals though; one for the size and then another for the "decrypting" state.